### DISCUSSION TITLE
data representations

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#12234] [Using MicroPython]
My tenuous grasp of upythons data formats & their representations is causing me grief again. Specifically the lack of symmetry when changing between formats keeps catching me out. 

At the moment I can put 2 x 8 bit analog values into a pair of bytes
```
>>> bytes([129])+bytes([20])
b'\x81\x14'
```
I can get the integer value
```
>>> struct.unpack('H',b'\x81\x14')[0]
5249
```
but when I try to backtrack to those original 2 values of 129 & 20
```
>>> hex(5249)
'0x1481'
```
I fall at the first hurdle with an '0x1481 representation instead of b'\x81\x14'

Can someone show me how to get from 5249 back to the 129 & 20 values I started with?

**Comment by robert-hh** [comment]
If you started with the  numbers 129 and 20, you can use arithmetic to get to 5249 as well by `256 * 20 + 129`, instead of using bytes() and struct(). From 5249 you can get back to 129 and 20 by `5249 & 0xff` and `5249 >> 8` (or `5249 % 256` and `5249 // 256`.
Obviously you can as well use `a=struct.pack("H", 5249)` to put 5249 into a bytes object `b"\x81\x14"` and use a[0] and a[1] to get the numbers 20 and 129.

**Comment by jimmo** [comment]
>  upythons data formats 

Just to clarify, this is standard Python behaviour, not MicroPython-specific.

> `bytes([129])+bytes([20])`

This is equivalent to `struct.pack('BB', 129, 20)` so if you want to unpack it, then you can use:

```py
x = struct.pack('BB', 129, 20) # x = b'\x81\x14'
a, b = struct.unpack('BB', x) # a = 129, b = 20
```

but you're working with bytes, so you could also just use

```py
x = struct.pack('BB', 129, 20) # x = b'\x81\x14'
a = x[0] # a = 129
b = x[1] # b = 20
```

**Reply by jimmo** [reply]
I realise I didn't answer your question about going back from the integer representation (see Robert's answer)

But you can also use struct here:

```
x = struct.pack('H', 5249)
a, b = struct.unpack('BB', x)
```

**Reply by jimmo** [reply]
> Specifically the lack of symmetry when changing between formats

@kjm1102 Hopefully this example shows the symmetry:

```py
# encoding side
a = 129
b = 20
x = struct.pack('BB', a, b)
y, = struct.unpack('H', x)

# y is now your 16 bit value

# decoding side
x = struct.pack('H', y)
a, b = struct.unpack('BB', x)
```

**Comment by GitHubsSilverBullet** [comment]
> Can someone show me how to get from 5249 back to the 129 & 20 values I started with?

```py
>>> x = 5249
>>> divmod(x, 256)
(20, 129)
>>> 
```

**Comment by kjm1102** [comment]
Thnx muchly, always possible to get back to the starting point if you're a competent enough programmer.

**Comment by peterhinch** [comment]
Another - arguably more self-evident - approach is to use a bytearray:
```python
a = bytearray(2)  # Store 2 values
a[0] = 20
a[1] = 129
print(a[0], a[1])
```
